### PR TITLE
HPCC-21766 Optimize common case of getUniqueId* calls

### PR DESCRIPTION
**Pull request by jakesmith**
Avoid finding foreign allocator if no foreign node supplied.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

File: dali/base/dacoven.cpp [hunk 1]
```diff
@@ -1009,7 +1028,7 @@ DALI_UID getGlobalUniqueIds(unsigned num,SocketEndpoint *_foreignnode)
     if (foreignnode.port==0)
```
Comment (mckellyln) [inline dali/base/dacoven.cpp]: Could it ever be that this routine is called but the _foreignnode argument supplied is null or 0.0.0.0 ?

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
@jakesmith looks good

**Comment by jakesmith** [comment]
@richardkchapman - ready to merge.

File: dali/base/dacoven.cpp [hunk 2]
```diff
@@ -581,23 +586,30 @@ class CCovenServer: public CCovenBase
...
+
+        CDaliUidAllocator *uidAllocator = nullptr;
+
         SocketEndpoint foreignnode;
         if (_foreignnode&&!_foreignnode->isNull()) {
             foreignnode.set(*_foreignnode);
             if (foreignnode.port==0)
                 foreignnode.port = DALI_SERVER_PORT;
+
+            uidcrit.enter();
```
Comment (richardkchapman) [inline dali/base/dacoven.cpp]: Why not use a CriticalBlock?

Comment (jakesmith) [reply dali/base/dacoven.cpp]: no good reason, just that this was pre-existing code (just moved a bit).
Could change?

Comment (richardkchapman) [reply dali/base/dacoven.cpp]: If you are going to refactor may as well do it properly

**Comment by richardkchapman** [comment]
@jakesmith Back to you

**Comment by jakesmith** [comment]
@richardkchapman - please review new commit / let me know when to squash.

File: dali/base/dacoven.cpp [hunk 3]
```diff
@@ -344,6 +329,23 @@ CIArrayOf<CDaliUidAllocator> CCovenBase::foreginUidallocators;
...
+CDaliUidAllocator &CDaliUidAllocator::find(CIArrayOf<CDaliUidAllocator> &uidallocators,const SocketEndpoint &foreignnode)
+{
+    CriticalBlock block(CCovenBase::uidcrit);
+    ForEachItemIn(i,uidallocators) {
+        if (uidallocators.item(i).node.equals(foreignnode))
+            return uidallocators.item(i);
+    }
+    CDaliUidAllocator &ret = *new CDaliUidAllocator(foreignnode);
+    uidallocators.append(ret);
+    if (uidallocators.ordinality()>1) {
```
Comment (richardkchapman) [inline dali/base/dacoven.cpp]: Probably not a biggie but this could be out of the critblock

**Comment by richardkchapman** [comment]
@jakesmith One minor comment - fix if you feel like it, then please squash

File: dali/base/dacoven.cpp (hunk 3 above)
Comment (jakesmith) [reply dali/base/dacoven.cpp]: that highlights something else, the logic here was weird, it assumed that 1st added was 'local' and any others were foreign.
But with this change any call to find is looking for a foreign uidallocator, so this condition (>1), shouldn't be here either.
I'll change.

**Comment by jakesmith** [comment]
@richardkchapman - haven't squashed yet, if you just look at last commit and then I'll squash.

**Comment by richardkchapman** [comment]
@jakesmith Looks fine - please squash

**Comment by jakesmith** [comment]
@richardkchapman - squashed

**Comment by richardkchapman** [comment]
@ghalliday Fancy a quick look before I merge

**Comment by ghalliday** [comment]
@richardkchapman looks good to me

**Pull request merged**